uri.go

```diff
@@ -0,0 +1,459 @@
+package anonymization
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/base32"
+	"fmt"
+
+	"github.com/intuitivelabs/sipsp"
+)
+
+const (
+	// maximum size allowed for an SIP URI is 2KB; with padding this results in at most 4KB
+	maxBufSize int = 1 << 12
+	// padding character used in base32 encoding
+	pad rune = '-'
+)
+
+// static buffers for encryption/decryption
+var (
+	encryptBuf [maxBufSize]byte
+	decryptBuf [maxBufSize]byte
+	encodeBuf  [maxBufSize]byte
+	decodeBuf  [maxBufSize]byte
+)
+
+type BlockModeCipher struct {
+	IV        []byte
+	Key       []byte
+	Block     cipher.Block
+	Encrypter cipher.BlockMode
+	Decrypter cipher.BlockMode
+}
+
+type UriCBC struct {
+	// user part cipher (key SHOULD be different from host part cipher)
+	User BlockModeCipher
+	// host part cipher (key SHOULD be different from user part cipher)
+	Host BlockModeCipher
+}
+
+func (bm *BlockModeCipher) Init(iv, key []byte, block cipher.Block) {
+	bm.Key = key
+	bm.IV = iv
+	bm.Block = block
+	bm.Encrypter = cipher.NewCBCEncrypter(bm.Block, bm.IV)
+	bm.Decrypter = cipher.NewCBCDecrypter(bm.Block, bm.IV)
+}
+
+var (
+	// URI CBC cipher
+	uriCBC = UriCBC{}
+	//uriGCM = UriGCM{}
+)
+
+func NewEncoding() *base32.Encoding {
+	return base32.HexEncoding.WithPadding(pad)
+}
+
+func NewUriCBC(iv, userKey, hostKey []byte) *UriCBC {
+	if block, err := aes.NewCipher(userKey); err != nil {
+		panic(err)
+	} else {
+		uriCBC.User.Init(iv, userKey, block)
+	}
+	if block, err := aes.NewCipher(hostKey); err != nil {
+		panic(err)
+	} else {
+		uriCBC.Host.Init(iv, hostKey, block)
+	}
+	return &uriCBC
+}
+
+func EncryptBuf() []byte {
+	return encryptBuf[:]
+}
+
+func DecryptBuf() []byte {
+	return decryptBuf[:]
+}
+
+func EncodeBuf() []byte {
+	return encodeBuf[:]
+}
+
+func DecodeBuf() []byte {
+	return decodeBuf[:]
+}
+
+type AnonymURI sipsp.PsipURI
+
+func (uri AnonymURI) String(buf []byte) string {
+	scheme := string(buf[uri.Scheme.Offs : uri.Scheme.Offs+uri.Scheme.Len])
+	user := string(buf[uri.User.Offs : uri.User.Offs+uri.User.Len])
+	pass := string(buf[uri.Pass.Offs : uri.Pass.Offs+uri.Pass.Len])
+	host := string(buf[uri.Host.Offs : uri.Host.Offs+uri.Host.Len])
+	port := string(buf[uri.Port.Offs : uri.Port.Offs+uri.Port.Len])
+	params := string(buf[uri.Params.Offs : uri.Params.Offs+uri.Params.Len])
+	headers := string(buf[uri.Headers.Offs : uri.Headers.Offs+uri.Headers.Len])
+	s := scheme
```

Comment:
I guess String() is for debugging  or seldom use only so we don't care about performance.
If you do it would be better to pre-alloc the string (compute the len and make()) and then append() each component to it.

Reply:
I am removing the function for the time being.